### PR TITLE
Use a TrieMap to manage pendingCommandIds in trigger runners

### LINKED ISSUE #15376: Pending command IDs need to be maintained using a trie map
Currently, trigger runner's maintain a map of pending command IDs (i.e. UUIDs) using a variable of immutable maps (ref: [Runner.scala#L363](https://github.com/digital-asset/daml/blob/main/triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala#L363)).

This variable is updated by code in the methods:
- [useCommandId](https://github.com/digital-asset/daml/blob/main/triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala#L370-L374)
- [handleCommands](https://github.com/digital-asset/daml/blob/main/triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala#L386-L401)

As these methods are called from an async context and `pendingCommandIds` is acting as a shared mutable variable (with no access/update concurrency controls), we have a pattern here that strongly suggests that `pendingCommandIds` can be modified in a manner that breaks mutual exclusion.

To fix this concern, we should use a [trie map](https://www.scala-lang.org/api/2.13.3/scala/collection/concurrent/TrieMap.html) and modify the variable to be a `val`.

### PR DESCRIPTION
**Pull request by carlpulley-da**
- Fixes #15376

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 1]
```diff
@@ -360,17 +360,27 @@ class Runner(
...
-  private[this] var pendingCommandIds = Map.empty[UUID, SeenMsgs]
+  private[this] val pendingCommandIds = TrieMap.empty[UUID, SeenMsgs]
   private val transactionFilter =
     TransactionFilter(parties.readers.map(p => (p.unwrap, trigger.filters)).toMap)
 
   private[this] def logger = ContextualizedLogger get getClass
 
   // return whether uuid *was* present in pendingCommandIds
   private[this] def useCommandId(uuid: UUID, seeOne: SeenMsgs.One): Boolean = {
-    val newMap = alterF(pendingCommandIds, uuid)(_ map (_ see seeOne))
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:371]: Refactored this to clarify its intention

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 2]
```diff
@@ -822,29 +832,23 @@ object Runner extends StrictLogging {
...
   private sealed abstract class SeenMsgs {
     import Runner.{SeenMsgs => S}
-    def see(msg: S.One): Option[SeenMsgs] =
-      (this, msg).match2 {
-        case S.Completion => { case S.Transaction => None }
-        case S.Transaction => { case S.Completion => None }
-        case S.Neither => { case one => Some(one: SeenMsgs) }
-      }(fallback = Some(this))
+
+    def see(msg: S.One): Option[SeenMsgs] = {
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:840]: Refactored to clarify intention

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 3]
```diff
@@ -385,7 +397,7 @@ class Runner(
   @throws[RuntimeException]
   private def handleCommands(commands: Seq[Command]): (UUID, SubmitRequest) = {
     val commandUUID = UUID.randomUUID
-    pendingCommandIds += ((commandUUID, SeenMsgs.Neither))
+    discard(pendingCommandIds.addOne(commandUUID -> SeenMsgs.Neither))
```
Comment (remyhaemmerle-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:400]:
```suggestion
    pendingCommandIds.update(commandUUID, SeenMsgs.Neither)
```
* avoid building a tuple
* avoid discard

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 4]
```diff
@@ -822,29 +834,23 @@ object Runner extends StrictLogging {
...
-        case S.Transaction => { case S.Completion => None }
-        case S.Neither => { case one => Some(one: SeenMsgs) }
-      }(fallback = Some(this))
+
+    def see(msg: S.One): Option[SeenMsgs] = {
+      (this, msg) match {
+        case (S.Completion, S.Transaction) =>
+          None
+        case (S.Transaction, S.Completion) =>
+          None
```
Comment (remyhaemmerle-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:847]:
```suggestion
        case (S.Completion, S.Transaction | S.Completion) =>
          None
```

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:842]: :+1:

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 5]
```diff
@@ -360,17 +360,29 @@ class Runner(
...
-    newMap foreach { pendingCommandIds = _ }
-    newMap.isDefined
+    pendingCommandIds.get(uuid) match {
+      case None =>
+        false
+
+      case Some(seenOne) =>
+        seenOne.see(seeOne) match {
+          case Some(v) =>
+            discard(pendingCommandIds.addOne(uuid -> v))
```
Comment (remyhaemmerle-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:380]:
```suggestion
            pendingCommandIds.update(uuid, v)
```

**Review by remyhaemmerle-da — APPROVED** [review]
Nice. Thanks.

**Pull request merged**